### PR TITLE
Add Client Lib Guidelines and Exporter Guidelines

### LINKED ISSUE #379: Move remaining documentation in google docs here (exporter guidelines, client lib etc)
Discussed it quickly in IRC: Most people feel the exporter guidelines (https://docs.google.com/document/d/1JapuiRbp-XoyECgl2lPdxITrhm5IyCUq9iA_h6jp3OY/edit#heading=h.pml2liglk52z) should be in the main docs.

### PR DESCRIPTION
**Pull request by discordianfish**
This closes #379
It's pretty much 1:1 translated. The only thing different/missing is the TOC.
Feel free to push additional fixes etc to this branch.

### CODE REVIEW

**Comment by beorn7** [comment]
This looks good to me. But I could imagine @juliusv or @brian-brazil feel way more strongly about website affairs than I do.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+More advanced uses cases (such as proxying from another monitoring/instrumentation system) require writing a custom Collector. Someone may also want to write a "bridge" that takes a CollectorRegistry and produces data in a format a different monitoring/instrumentation system understands, allowing users to only have to think about one instrumentation system.
+
+CollectorRegistry SHOULD offer register()/unregister() functions, and a Collector SHOULD be allowed to be registered to multiple CollectorRegistrys.
+
+Client libraries MUST be thread safe.
+
+For non-OO languages such as C, client libraries should follow the spirit of this structure as much as is practical.
+
+Naming
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Should this be a heading?

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+These should be primarily used as file-static variables, that is, global variables defined in the same file as the code they’re instrumenting. The client library SHOULD enable this. The common use case is instrumenting a piece of code overall, not a piece of code in the context of one instance of an object. Users shouldn’t have to worry about plumbing their metrics throughout their code, the client library should do that for them (and if it doesn’t, users will write a wrapper around the library to make it "easier" - which rarely tends to go well).
+
+There MUST be a default CollectorRegistry, the standard metrics MUST by default implicitly register into it with no special work required by the user. There MUST be a way to have metrics not register to the default CollectorRegistry, for use in batch jobs and unittests. Custom collectors SHOULD also follow this.
+
+Exactly how the metrics should be created varies by language. For some (Java, Go) a builder approach is best, whereas for others (Python) function arguments are rich enough to do it in one call.
+
+For example in the Java Simpleclient we have:
+
+```
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Let's add `java` for code highlighting.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+This will register requests with the default CollectorRegistry. By calling build() rather than register() the metric won’t be registered (handy for unittests), you can also pass in a CollectorRegistry to register() (handy for batch jobs).
+
+### Counter
+
+Counter is a monotonically increasing counter. It MUST NOT allow the value to decrease, however it MAY be reset to 0 (such as by server restart).
+
+A counter MUST have the following methods:
+
+inc(): Increment the counter by 1
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Let's mark code as code with backticks `inc()`.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+Counters MUST start at 0.
+
+### Gauge
+
+Gauge represents a value that can go up and down.
+
+A gauge MUST have the following methods:
+
+inc(): Increment the gauge by 1
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Looks like these methods are supposed to be a list.

**Comment by discordianfish** [comment]
Okay then.. looks like if I don't do the reformatting nobody will ;).
Thrown in some `` and - to make things look nicer.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
+---
+title: Client Library Guidelines
+sort_rank: 2
+---
+
+# Client Library Guidelines
+
+This document covers what functionality and API Prometheus client libraries should offer, with the aim of consistency across libraries, making the easy use cases easy and avoiding offering functionality that may lead users down the wrong path.
+
+There are 4 languages supported at the time of writing, so we’ve gotten a good sense by now of how to write a client. These guidelines aim to help authors of new client libraries produce good libraries.
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: `s;4 languages supported;[10 languages already supported](/docs/instrumenting/clientlibs);`

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+## Conventions
+
+MUST/MUST NOT/SHOULD/SHOULD NOT/MAY have the meanings given in [https://www.ietf.org/rfc/rfc2119.txt](https://www.ietf.org/rfc/rfc2119.txt)
+
+In addition ENCOURAGED means that a feature is desirable for a library to have, but it’s okay if it’s not present. In other words, a nice to have.
+
+Things to keep in mind
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Add colon for the following list.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+* Take advantage of each language’s features
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: End all these sentences with a full stop.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+### Naming
+
+Client libraries SHOULD follow function/method/class names mentioned in this document, keeping in mind the naming conventions of the language they’re working in. For example, `set_to_current_time()` is good for a method name Python, but `SetToCurrentTime()` is better in Go and `setToCurrentTime()` is the convention in Java. Where names differ for technical reasons (e.g. not allowing function overloading), documentation/help strings SHOULD point users towards the other names.
+
+Libraries MUST NOT offer functions/methods/classes with the same or similar names to ones given here, but with different semantics.
+
+## Standard metrics types
+
+Counter, Gauge, Summary and Histogram are the primary interface by users.
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Maybe `The Counter, Gauge, Summary and Histogram [Metric types](/docs/concepts/metric_types/) are the primary interface by users.`

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+A counter is ENCOURAGED to have:
+
+A way to count exceptions throw/raised in a given piece of code, and optionally only certain types of exceptions. This is count_exceptions in Python.
+
+Counters MUST start at 0.
+
+### Gauge
+
+Gauge represents a value that can go up and down.
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Let's link all the metric names to the respective page on https://prometheus.io/docs/concepts/metric_types/. E.g. `[Gauge](/docs/concepts/metric_types/#gauge)`.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+**Further metrics considerations**
+
+Providing additional functionality in metrics beyond what’s documented above as makes sense for a given language is ENCOURAGED. 
+
+If there’s a common use case you can make simpler then go for it, as long as it won’t encourage undesirable behaviours (such as suboptimal metric/label layouts, or doing computation in the client).
+
+### Labels
+
+Labels are one of the most powerful aspects of Prometheus, but easily abused. Accordingly client libraries must be very careful in how labels are offered to users. See [http://prometheus.io/docs/practices/instrumentation/#use-labels](http://prometheus.io/docs/practices/instrumentation/#use-labels) and [http://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels](http://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels)
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: > Labels are one of the [most powerful aspects](/docs/practices/instrumentation/#use-labels) of Prometheus, but [easily abused](/docs/practices/instrumentation/#do-not-overuse-labels). Accordingly client libraries must be very careful in how labels are offered to users.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+Client libraries MUST NOT under any circumstances allow users to have different label names for the same metric for Gauge/Counter/Summary/Histogram or any other Collector offered by the library.
+
+If your client library does validation of metrics at collect time, it MAY also verify this for custom Collectors.
+
+While labels are powerful, the majority of metrics will not have labels. Accordingly the API should allow for labels but not dominate it.
+
+A client library MUST allow for optionally specifying a list of label names at Gauge/Counter/Summary/Histogram creation time. A client library SHOULD support any number of label names. A client library MUST validate that label names meet the requirements in [http://prometheus.io/docs/concepts/data_model/#metric-names-and-labels](http://prometheus.io/docs/concepts/data_model/#metric-names-and-labels)
```
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: > ... meet the [documented requirements](/docs/concepts/data_model/#metric-names-and-labels).

File: content/docs/instrumenting/exporter_guidelines.md [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+sort_rank: 5
+---
+
+# Exporter Guidelines
+
+When directly instrumenting your own code, the general rules of how to instrument code with a Prometheus client library can be followed quite directly. When taking metrics from another monitoring or instrumentation system, things tend not to be so black and white.
+
+This document contains things you should consider when writing an exporter or custom collector. The theory covered will also be of interest to those doing direct instrumentation.
+
+If you are writing an exporter and are unclear on anything here, contact us on IRC (#prometheus on Freenode) or the mailing list.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: `[mailing list](/community)`

File: content/docs/instrumenting/exporter_guidelines.md [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+
+In the more common case metrics from the system are completely non-standard, depending on how the user is using it and what the underlying application is. In that case the user has to tell us how to transform the metrics. The JMX exporter is the worst offender here, with the graphite and statsd exporters also requiring configuration to extract labels.
+
+Providing something that produces some output out of the box and a selection of example configurations is advised. When writing configurations for such exporters, this document should be kept in mind.
+
+YAML is the standard Prometheus configuration format.
+
+## Metric Naming
+
+[http://prometheus.io/docs/practices/naming/](http://prometheus.io/docs/practices/naming/) has the general rules on metric naming.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: > Follow the [best practices on metric naming](/docs/practices/naming).

File: content/docs/instrumenting/exporter_guidelines.md [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+Metric names should not include the labels that they’re exported with (e.g. by_type) as that won’t make sense if the label is aggregated away.
+
+The one exception is when you’re exporting the same data with different labels via multiple metrics, in which case that’s usually the sanest way to distinguish them. For direct instrumentation this should only come up when exporting a single metric with all the labels would have too high a cardinality.
+
+Prometheus metrics and label names are written in snake_case. Converting camelCase to snake_case is desirable, though it doing so automatically doesn’t always produce nice results for things like myTCPExample or isNaN so sometimes it’s best to leave them as-is.
+
+Exposed metrics should not contain colons, these are for users to use when aggregating.
+
+Only [a-zA-Z0-9:_] are valid in metric names, any other characters should be sanitized to an underscore.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Use backticks for the regular expression.

File: content/docs/instrumenting/exporter_guidelines.md [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+
+[http://prometheus.io/docs/practices/naming/](http://prometheus.io/docs/practices/naming/) has the general rules on metric naming.
+
+Generally metric names should allow someone who’s familiar with Prometheus but not a particular system to make a good guess as to what a metric means.  A metric named ‘http_requests_total’ is not extremely useful - are these being measured as they come in, in some filter or when they get to the user’s code? ‘requests_total’ is even worse, what type of requests?
+
+To put it another way with direct instrumentation, a given metric should exist within exactly one file. Accordingly within exporters and collectors, a metric should apply to exactly one subsystem and be named accordingly.
+
+Metric names should never be procedurally generated, except when writing a custom collector or exporter.
+
+Metric names for applications should generally be prefixed by the exporter name, e.g. haproxy_up.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Use backticks around the metric name.

File: content/docs/instrumenting/exporter_guidelines.md [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+
+Metrics must use base units (e.g. seconds, bytes) and leave converting them to something more readable to the graphing software. No matter what units you end up using, the units in the metric name must match the units in use. Similarly expose ratios, not percentages (though a counter for each of the two components of the ratio is better).
+
+Metric names should not include the labels that they’re exported with (e.g. by_type) as that won’t make sense if the label is aggregated away.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Use backticks around `by_type`.

File: content/docs/instrumenting/exporter_guidelines.md [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+
+The one exception is when you’re exporting the same data with different labels via multiple metrics, in which case that’s usually the sanest way to distinguish them. For direct instrumentation this should only come up when exporting a single metric with all the labels would have too high a cardinality.
+
+Prometheus metrics and label names are written in snake_case. Converting camelCase to snake_case is desirable, though it doing so automatically doesn’t always produce nice results for things like myTCPExample or isNaN so sometimes it’s best to leave them as-is.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Backticks around `snack_case` and `camelCase`.

File: content/docs/instrumenting/exporter_guidelines.md [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+
+Exposed metrics should not contain colons, these are for users to use when aggregating.
+
+Only [a-zA-Z0-9:_] are valid in metric names, any other characters should be sanitized to an underscore.
+
+The _sum, _count, _bucket and _total suffixes are used by Summaries, Histograms and Counters. Unless you’re producing one of those, avoid these suffixes.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Backticks around `_sum`, `_bucket` and `_total`.

File: content/docs/instrumenting/exporter_guidelines.md [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+
+Exposed metrics should not contain colons, these are for users to use when aggregating.
+
+Only [a-zA-Z0-9:_] are valid in metric names, any other characters should be sanitized to an underscore.
+
+The _sum, _count, _bucket and _total suffixes are used by Summaries, Histograms and Counters. Unless you’re producing one of those, avoid these suffixes. 
+
+_total is a convention for counters, you should use it if you’re using the COUNTER type.
+
+The process_ and scrape_ prefixes are reserved. It’s okay to add your own prefix on to these if they follow the matching semantics. E.g. Prometheus has scrape_duration_seconds for how long a scrape took, it’s good practice to have e.g. jmx_scrape_duration_seconds saying how long the JMX collector took to do it’s thing. For process stats where you have access to the pid, both Go and Python offer collectors that’ll handle this for you (see the haproxy exporter for an example).
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Backticks around everything which is a metric.

File: content/docs/instrumenting/exporter_guidelines.md [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+
+When you have a successful request count and a failed request count, the best way to expose this is as one metric for total requests and another metric for failed requests. This makes it easy to calculate the failure ratio. Do not use one metric with a failed/success label. Similarly with hit/miss for caches, it’s better to have one metric for total and another for hits.
+
+Consider the likelihood that someone using monitoring will do a code or web search for the metric name. If the names are very well established and unlikely to be used outside of the realm of people used to those names (e.g. SNMP and network engineers) then leaving them as-is may be a good idea. This logic doesn’t apply for e.g. MySQL as non-DBAs can be expected to be poking around the metrics. A HELP string with the original name can provide most of the same benefits as using the original names.
+
+## Labels
+
+[http://prometheus.io/docs/practices/instrumentation/#things-to-watch-out-for](http://prometheus.io/docs/practices/instrumentation/#things-to-watch-out-for) has general advice on labels.
```
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: > Read the [general advice](/docs/practices/instrumentation/#things-to-watch-out-for) on labels.

**Comment by grobie** [comment]
I made all the changes and more. Please review @discordianfish, @brian-brazil and @juliusv.

**Comment by discordianfish** [comment]
Looks good to me, but now that we have content changes we definitely need to get @brian-brazil to review it as well.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 21]
```diff
@@ -152,71 +152,71 @@ A histogram MUST have the following methods:
...
 
 The Child returned by `labels()` SHOULD be cacheable by the user, to avoid having to look it up again - this matters in latency-critical code.
 
 Metrics with labels SHOULD support a `remove()` method with the same signature as `labels()` that will remove a Child from the metric no longer exporting it, and a `clear()` method that removes all Children from the metric. These invalidate caching of Children.
 
-There SHOULD be a way to initialize a given Child with the default value, usually just calling `labels()`. No-labels metrics MUST always be initialized. See [http://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics](http://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics)
+There SHOULD be a way to initialize a given Child with the default value, usually just calling `labels()`. Metrics without labels MUST always be initialized to avoid [problems with missing metrics](/docs/practices/instrumentation/#avoid-missing-metrics).
 
-## **Metric names**
+### Naming
```
Comment (brian-brazil) [inline content/docs/instrumenting/clientlib_guidelines.md]: The original heading of "metric names" is correct here, as this section is just about metric names. Naming of other things such as the API is covered elsewhere

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 22]
```diff
@@ -152,71 +152,71 @@ A histogram MUST have the following methods:
...
-Metric names must match the restrictions in [http://prometheus.io/docs/concepts/data_model/#metric-names-and-labels](http://prometheus.io/docs/concepts/data_model/#metric-names-and-labels). As with label names, this MUST be met for uses of Gauge/Counter/Summary/Histogram and in any other Collector offered with the library.
+Metric names must follow the [specification](/docs/concepts/data_model/#metric-names-and-labels). As with label names, this MUST be met for uses of Gauge/Counter/Summary/Histogram and in any other Collector offered with the library.
 
-Many client libraries offer setting the name in three parts: namespace_subsystem_name of which only the "name" is mandatory.
+Many client libraries offer setting the name in three parts: `namespace_subsystem_name` of which only the `name` is mandatory.
 
 Dynamic/generated metric names or subparts of metric names MUST be discouraged, except when a custom Collector is proxying from other instrumentation/monitoring systems. Generated/dynamic metric names are a sign that you should be using labels instead.
 
-## **Metric descriptions/help**
+### Description/help
```
Comment (brian-brazil) [inline content/docs/instrumenting/clientlib_guidelines.md]: Ditto, this is about metrics

File: content/docs/instrumenting/exporter_guidelines.md [hunk 23]
```diff
@@ -112,55 +114,69 @@ The first is some label it’d be useful to have on the metrics that are about,
 
 The other case are what are really target labels. These are things like region, cluster names, and so on, that come from your infrastructure setup rather than the application itself. It’s not for an application to say where it fits in your label taxonomy, that’s for the person running the Prometheus server to configure and different people monitoring the same application may give it different names.
 
-Accordingly these labels belong up in the scrape configs of Prometheus via whatever service discovery you’re using. It’s okay to apply [http://www.robustperception.io/how-to-have-labels-for-machine-roles/](http://www.robustperception.io/how-to-have-labels-for-machine-roles/) here too, as it’s likely useful information for at least some of the people scraping it.
+Accordingly these labels belong up in the scrape configs of Prometheus via whatever service discovery you’re using. It’s okay to apply the linked principe here too, as it’s likely useful information for at least some of the people scraping it.
```
Comment (brian-brazil) [inline content/docs/instrumenting/exporter_guidelines.md]: principle, and link it

File: content/docs/instrumenting/exporter_guidelines.md [hunk 24]
```diff
@@ -194,35 +210,21 @@ The second is that these sort of systems tend to allow your users to send either
...
-In the Java world, many instrumentation frameworks expose process-level and JVM-level stats such as CPU and GC. The Java client and JMX exporter already include these in the preferred form via DefaultExports.java, so these should be dropped.
-
-Similarly with other languages.
+The latter is better where there’s still some useful metrics you can get even with a failed scrape, such as the haproxy exporter providing process stats. The former is a tad easier for users to deal with, as `up` works in the usual way (though you can’t distinguish between the exporter being down and the application being down).
 
-## Landing page
+### Landing page
 
-It’s nicer for users if visiting [http://yourexporter/](http://yourexporter/) has a simple html page with the name of the exporter, and a link to the /metrics.
+It’s nicer for users if visiting [http://yourexporter/](http://yourexporter/) has a simple html page with the name of the exporter, and a link to the `/metrics`.
```
Comment (brian-brazil) [inline content/docs/instrumenting/exporter_guidelines.md]: this probably shouldn't be a link

File: content/docs/instrumenting/exporter_guidelines.md [hunk 25]
```diff
@@ -234,4 +236,4 @@ This is a registry to make our users’ lives a little easier, not a commitment
 
 ## Announcing
 
-Once you’re ready to announce your exporter to the world, send an email to the mailing list and send a PR to add it to [http://prometheus.io/docs/instrumenting/exporters/](http://prometheus.io/docs/instrumenting/exporters/)
+Once you’re ready to announce your exporter to the world, send an email to the mailing list and send a PR to add it to [the list of available exporters](/docs/instrumenting/exporters).
```
Comment (brian-brazil) [inline content/docs/instrumenting/exporter_guidelines.md]: should we link to the github file here?

File: content/docs/instrumenting/clientlib_guidelines.md (hunk 21 above)
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Yes. This is a subheading of `Metrics`.

File: content/docs/instrumenting/clientlib_guidelines.md (hunk 22 above)
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Yupp, that's why its part of the Metrics section.

File: content/docs/instrumenting/exporter_guidelines.md (hunk 23 above)
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: It's already linked in the same section just two sentences above. Following the wikipedia principle, only the first occurrence gets linked. I'll try to rewrite the sentence; I had to read the full thing a couple of times until I started to understand what it is about.

File: content/docs/instrumenting/exporter_guidelines.md (hunk 24 above)
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Makes sense, removed.

File: content/docs/instrumenting/exporter_guidelines.md (hunk 25 above)
Comment (grobie) [inline content/docs/instrumenting/exporter_guidelines.md]: Makes sense. Done.

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 26]
```diff
@@ -0,0 +1,249 @@
+---
+title: Client Library Guidelines
```
Comment (juliusv) [inline content/docs/instrumenting/clientlib_guidelines.md]: No title case in menu items ("Client library guidelines")

File: content/docs/instrumenting/clientlib_guidelines.md [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+sort_rank: 2
+---
+
+# Client Library Guidelines
+
+This document covers what functionality and API Prometheus client libraries should offer, with the aim of consistency across libraries, making the easy use cases easy and avoiding offering functionality that may lead users down the wrong path.
```
Comment (juliusv) [inline content/docs/instrumenting/clientlib_guidelines.md]: We wrap lines at around 80 chars everywhere else in the docs, would be good to do the same here.

File: content/docs/instrumenting/clientlib_guidelines.md (hunk 21 above)
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Alright, it's not clear in links though, as the anchor would only be `...#naming`. I changed it back.

File: content/docs/instrumenting/clientlib_guidelines.md (hunk 22 above)
Comment (grobie) [inline content/docs/instrumenting/clientlib_guidelines.md]: Change it back.

**Comment by juliusv** [comment]
In general, I see a lot of style and OCD kind of things I'll still want to change about this. But I won't be able to look at this anymore today, so feel free to merge for now and I'll fix it up later.

**Comment by grobie** [comment]
Alright, I made all changes requested by Brian. Merging for now as I believe it's already easier to read than the google doc. Let's iterate on it.

**Pull request merged**